### ISSUE TITLE
Vavr arguments do not work with SqlObjects

### ISSUE THREAD

**Issue by diversit** [jdbi/jdbi#2529]
Using Vavr object like an `Option` as an argument with SqlObjects like
```java
@SqlQuery("select * from something where :name is null or name = :name order by id")
@RegisterBeanMapper(Something.class)
List<Something> selectByOptionName(@Bind Option<String> name);
```
does not work.

Based on the `TestVavrValueArgumentFactoryWithDB` in JDBI Vavr module, I created these test cases for using an Option as an argument. 
```java
public class TestVavrValueArgumentFactoryWithDBSqlObjects {


    private static final Something ERICSOMETHING = new Something(1, "eric");
    private static final Something BRIANSOMETHING = new Something(2, "brian");

    @RegisterExtension
    public JdbiExtension h2Extension = JdbiExtension.h2().withInitializer(TestingInitializers.something()).installPlugins();

    @BeforeEach
    public void createTestData() {
        Handle handle = h2Extension.getSharedHandle();
        handle.createUpdate("insert into something (id, name) values (1, 'eric')").execute();
        handle.createUpdate("insert into something (id, name) values (2, 'brian')").execute();
    }


    @Test
    public void testGetOptionShouldReturnCorrectRowUsingExtension() {
        List<Something> result = h2Extension.getJdbi()
                .withExtension(Dao.class, dao -> dao.selectByOptionName(Option.of("eric")));

        assertThat(result).hasSize(1);
    }

    @Test
    public void testGetOptionEmptyShouldReturnAllRowsUsingExtension() {
        List<Something> result = h2Extension.getJdbi()
                .withExtension(Dao.class, dao -> dao.selectByOptionName(Option.none()));

        assertThat(result).hasSize(2);
    }

    interface Dao {

        @SqlQuery("select * from something where :name is null or name = :name order by id")
        @RegisterBeanMapper(Something.class)
        List<Something> selectByOptionName(@Bind Option<String> name);
    }
```

With a `Some` the query works, but with a `None` it does not.
This is somewhat different than what we see in our codebase. Then the test with a `Some` value also fails with 
```
org.jdbi.v3.core.statement.UnableToExecuteStatementException: org.postgresql.util.PSQLException: ERROR: operator does not exist: integer = bigint[]
  Hint: No operator matches the given name and argument types. You might need to add explicit type casts.
  Position: 50 [statement:"SELECT * FROM our_table  WHERE id = :id ", arguments:{positional:{0:Some(47)}, named:{id:Some(47)}, finder:[]}]
```

While debugging I notice that the `VavrValueArgumentFactory.build` method does not get called when using SqlObject.
It _does_ get called when using the Java Api `.bind("name", Option.none())`.

**Comment by hgschmie** [comment]
Hi @diversit,

Thank you for opening a bug report with the Jdbi team. I can reproduce that problem locally (also the Postgres errors).

**Comment by hgschmie** [comment]
The problem is that `Option` implements `Iterable` and when creating the arugment, the SqlArrayArgumentFactory "snatches" the argument because it deals with all iterable arguments. This is a shortcoming in factory selection. I'll figure something out.

**Comment by hgschmie** [comment]
Will be fixed in the next release (3.42.0)
